### PR TITLE
generic loading component

### LINKED ISSUE #3587: A generic and awesome lazy-load fallback component
I know I'm going off tangent here a bit but it's worth discussing. We should probably have 1 single useful component for all SPAs that might need it. And it should be smart and re-usable. 
In particular, I don't think it should immediately say `<p>Loading...</p>` but rather only say that after a while. E.g.

```jsx
function Loading({message, delay, minHeight} = {message: "Loading...", delay: 500, minHeight: 200}: {message: string, delay: number}) {
  ...
}
```
this component should display an empty string for about 500ms and only if more than 500ms has passed, do we display the `Loading...` text. I don't know if 500ms is correct, but the idea is that most people on a decent connection will get their lazy-loaded stuff in a matter of less than 1 second (thank you Cloudfront CDN!) so it would just be annoying to display "Loading..." and then almost immediately (from a human's eyes point of view) replacing with something else. Then it's much better to stare at a blank white if the lazy-loaded app comes in sufficiently fast. 

The places where we can immediately benefit from a component like this are:

1. Lazy-loading the BCD table ([current implementation](https://github.com/mdn/yari/blob/ce67d66ecd57752801745e9e4d77944925a85c5e/client/src/document/lazy-bcd-table.tsx#L89-L95))
2. The site-search search results page
3. Sign-in and sign-up (when it waits for `/api/v1/whoami`)
4. Account settings page
5. Client-side navigation/loading of documents. E.g. when you click a title-autocomplete search result or click a breadcrumb link.

I think it's important to build this component without assuming it can take over the whole page. It needs to be possible to put in without it needing to have classNames set on the containing parent. 

And by the way, this component should probably have some decent default `min-height` which you can override. E.g. default of 200px but we can tune that based on afterward observing how high the lazy-loaded content is when it comes in.

_Originally posted by @peterbe in https://github.com/mdn/yari/issues/3584#issuecomment-824335496_

### PR DESCRIPTION
**Pull request by peterbe**
Fixes #3587

The general idea is two-fold:

1. Unless the loading is really slow (or almost stuck) is there any point in showing a loading message. Blank whiteness is good enough and affording. Most lazy-loading or XHR loading is actually pretty fast. Very often <=1 sec. So it's just detrimental to show a flickering "Loading" for a couple of milliseconds. 
2. One generic solution for all of these that we can use for all sorts, from code-split lazy-loading, to waiting for XHR requests. 

@schalkneethling Clearly, I don't understand how that `@include fade-in();` works. Can you please advice. 
Also, I think it would be nice if (instead) the loading message could gently and slowly pulsate. So maybe that `slow-pulse` mixin is better. Because what I put in there isn't really working, I'm going to make this a draft.

### CODE REVIEW

**Comment by peterbe** [comment]
Also, I believe there are some more places where we're *going* to want to use this once it's in `main`. E.g. the landing page survey component loading.

**Comment by peterbe** [comment]
REmember, if your lazy-loading is too fast; https://www.peterbe.com/plog/how-to-simulate-slow-lazy-chunk-loading-in-react

**Comment by schalkneethling** [comment]
@peterbe I guess this will then also take care off https://github.com/mdn/yari/issues/3341?

**Comment by peterbe** [comment]
> @peterbe I guess this will then also take care of #3341?

Yikes. Yeah, you're right. It's not the first time I've filed a duplicate without doing a search first. But I think my "excuse" was that I started with the code, and worked backwards. I.e. create the code, *then* write up an issue to go with the incoming PR. Hence the poor research before filing.

**Comment by peterbe** [comment]
Sorry. Didn't actually mean to close. Fumbled on the button.

**Comment by peterbe** [comment]
Incidentally, I actually tackled https://github.com/mdn/yari/issues/2990 too with this PR. 
So now, document loading gets the same generic loading fallback but no faint dino head. 
I quite liked the dino as a visual thing. Perhaps we should add that. But I certainly like the bonus that document loading gets the same component as any other page or app.

**Comment by peterbe** [comment]
@schalkneethling So, can you help me get that fade-in and smooth-pulse added to this PR?

**Comment by schalkneethling** [comment]
> @schalkneethling So, can you help me get that fade-in and smooth-pulse added to this PR?

Will do.

File: client/src/ui/atoms/loading/index.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * Use setTimeout with Hooks in a declarative way.
+ *
+ * @see https://stackoverflow.com/a/59274757/3723993
+ * @see https://overreacted.io/making-setinterval-declarative-with-react-hooks/
+ */
+export function useTimeout(
+  callback: React.EffectCallback,
+  delay: number | null
```
[the remark this answers is not in the document]
Comment (peterbe) [reply client/src/ui/atoms/loading/index.tsx:13]: Gee. I copy-n-pasted this (probably from Dan's blog post) so long ago that I've forgotten about it.

**Comment by peterbe** [comment]
@schalkneethling Whilst I tackle Gregor's feedback, do you think you can make a cross-PR to implement the SCSS animations properly. If you think it's still a good idea to slowly/lazily load in the dino background, we can perhaps do that some other day as soon as this is merged.

**Comment by schalkneethling** [comment]
> @schalkneethling Whilst I tackle Gregor's feedback, do you think you can make a cross-PR to implement the SCSS animations properly. If you think it's still a good idea to slowly/lazily load in the dino background, we can perhaps do that some other day as soon as this is merged.

Will do.

**Comment by peterbe** [comment]
@Gregoor I took all your review feedback. Seems to work. 

@schalkneethling As you review this and hack on the `fade-in` and `slow-pulse`, can you please also keep an eye on the `minHeight` I used in various places. 
I know it has a default of `200px` but I wonder if that's useful. Perhaps it's wrong to even have a default. 
Perhaps we should "carefully" review every instance of this being used and have a think about what the appropriate `minHeight` should be in each individual case. You can never predict exactly what height it will become but ultimately, we're doing this so that the black footer isn't appearing in the viewport, just to disappear 1 second later causing a flicker from black to white.

File: client/src/document/index.tsx [hunk 2]
```diff
@@ -132,7 +131,8 @@ export function Document(props /* TODO: define a TS interface for this */) {
   }, []);
 
   if (!doc && !error) {
-    return <LoadingDocumentPlaceholder />;
+    // return <LoadingDocumentPlaceholder />;
```
Comment (schalkneethling) [inline client/src/document/index.tsx:134]: I guess we can remove this?

**Comment by schalkneethling** [comment]
@peterbe https://github.com/peterbe/yari/pull/11

**Review by schalkneethling — APPROVED** [review]
Other than the one commented-out line that can probably be deleted, this is all good. Gonna miss the dino though :)

**Comment by escattone** [comment]
> Other than the one commented-out line that can probably be deleted, this is all good. Gonna miss the dino though :)

@schalkneethling Me too! I liked the dino.

**Review by schalkneethling — APPROVED** [review]

**Comment by peterbe** [comment]
@schalkneethling 
Perhaps it's something I've accidentally set on my macOS so it doesn't do any animations. 
In Chrome:
https://user-images.githubusercontent.com/26739/119696619-630c1f80-be1d-11eb-85ff-48b2955673d9.mov

In Firefox Nightly:

https://user-images.githubusercontent.com/26739/119696775-82a34800-be1d-11eb-833f-9bc963cc5843.mov

I used https://www.peterbe.com/plog/how-to-simulate-slow-lazy-chunk-loading-in-react to slow down the loading of that `app.tsx` in the `settings/index.tsx`. 
Any idea?

**Comment by schalkneethling** [comment]
> Any idea?

Prefers reduced motion? I think that is under accessibility settings on macOS

Screenshot 2021-05-26 at 18 33 59

**Comment by peterbe** [comment]
> > Other than the one commented-out line that can probably be deleted, this is all good. Gonna miss the dino though :)
> 
> @schalkneethling Me too! I liked the dino.

I deliberately didn't delete the `dino.svg` icon from the `assets` folder. Hint hint. 
I think it would be pretty if that `dino.svg` could load as a backdrop to the "Loading..." message. (let's just not include it in React :)

**Comment by escattone** [comment]
> > > Other than the one commented-out line that can probably be deleted, this is all good. Gonna miss the dino though :)
> > 
> > 
> > @schalkneethling Me too! I liked the dino.
> 
> I deliberately didn't delete the `dino.svg` icon from the `assets` folder. Hint hint.
> I think it would be pretty if that `dino.svg` could load as a backdrop to the "Loading..." message. (let's just not include it in React :)

@schalkneethling @peterbe I really like the idea of having the dino in there as a background. Let's do that if it works for you guys.

**Pull request merged**